### DISCUSSION TITLE
Deleting protected branches is hard

### DISCUSSION THREAD

**Discussion by clphillips** [community/community#21793] [New to GitHub]
<p>Not long ago I requested <a href="https://github.community/t/branch-protection-on-multiple-branches/1198/6">Branch Protection on multiple branches</a>, and that works great (thanks github!). However, there is one fatal flaw with it:</p>
<p>Removing stale protected branches is really hard.</p>
<p>For example, say my branch protection pattern is release-?.?, once I’ve released release-4.0 I want to delete release-3.0, but in order to do so, I have to remove my branch protection, delete it, then quickly add the branch protection back.</p>
<p>It would be great if I could delete a protected branch as an admin with some kind of confirmation.</p>

**Comment by mpboom** [comment]
<p>Why not leave the protected branch up on GitHub? It’s always nice to have some kind of back-up to see where things come from.</p>

**Comment by clphillips** [comment]
<p>We have tags for backup. The branch is just added noise and can confuse engineers creating new PRs.</p>

**Comment by The-Futurist** [comment]
<p>This is interesting and you’re being impacted as we are too (but for a different reason). Please see my recent post:</p>
<p><a href="https://github.community/t/code-review-problems/1476">https://github.community/t5/How-to-use-Git-and-GitHub/Code-Review-Problems/td-p/14439</a></p>

**Comment by outdooricon** [comment]
<p>I agree, this is a problem for us too. A lot of our protected branches only need to exist for a time, and then we remove them. It’d be nice to at least have admin delete rights on the branches list</p>

**Comment by szekely-szabolcs** [comment]
<p>The issue affects us, as well.</p>
<p>Admins should have the right to delete protected branches, maybe with extra confirmation checkbox, like “I want to use my admin priviledges…”</p>

**Comment by ChocPanda** [comment]
<p>We have a similar problem, we were trying to follow a similar development pattern as described <a href="https://stackoverflow.com/questions/32024344/git-model-for-story-branch-with-many-feature-branches" rel="nofollow noopener">here on stackoverflow</a> and introduced branch protection and hooks for the story branches however they can’t be removed once the story is complete and merged in master.</p>

**Comment by dominik-pakosz** [comment]
<p>Agree.<br>
I have protection on <code>feature/*</code> branches. Sometimes feature fail during development and I want to delete this branch. Can’t do this. It there any solution for deleting them?</p>

**Comment by leotm** [comment]
<p>I just temporarily amended the <em>Branch protection rule</em> so that it applies to <em>0</em> branches, removed the desired branches, then reverted the rule.</p>

**Comment by EGHornbostel** [comment]
<p>I’ve done this too. It <em>works</em> but it’s not ideal</p>

**Comment by ThirdEyeRose** [comment]
<p>Hello <a class="mention" href="/clphillips">@clphillips</a> and everyone else in this thread!</p>
<p>I’m very sorry for the delay, but I wanted to let you know that I’ve compiled all of this great feedback into a feature request and have sent it off to the product team. I can’t say if or when this feature might be implemented, but when it is implemented, it will be announced in the <a href="https://github.blog/changelog/" rel="nofollow noopener">GitHub Changelog</a>.</p>
<p>Thanks!</p>

**Comment by clakech** [comment]
<p>Hi,</p>
<p>We also need to be able to delete protected branches without being a developer for “feature/*” branches for instance that once merged into master cannot be deleted easily.</p>
<p>We have to ask to an admin (me) and this doesn’t scale way or we have to change the branch policy, delete the branch, then re-set the branch policy again… this is annoying right ?</p>

**Comment by RdeWilde** [comment]
<p>Enabling deleting branches for administrators, overruling it with a checkbox like PR’s can, would be a big help</p>

**Comment by theomessin** [comment]
<p>Any update on this?</p>
<p>I don’t see why delete protection must be a default for all branch protections. Can you not just have it as an extra tick box and let us decide whether we need this functionality?</p>
<p>Eg. we might want to use branch protections only to enforce status checks for feature branches that need to be deleted after a PR. Currently, not even GitHub can delete those branches following a merged PR even if the  <strong>Automatically delete head branches</strong>  option is selected.</p>

**Comment by kachow6** [comment, marked as answer]
<p>For those who join this thread late or who have been waiting, I checked my branch protection rules today and found the feature implemented!</p>
<p>You can find the new force push and branch deletion rules at the bottom of the branch protection rule page for the branch name pattern you would like to apply rule to:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/1X/4f5590e1d27491215bc4f1ba99df271f762af756.png" data-download-href="/uploads/short-url/bjP2KvWFalPHOosA5UarSfw12Oq.png?dl=1" title="Screen Shot 2019-12-04 at 4.51.41 PM.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/25444249/181066911-6e50942f-b784-40cb-877b-c80e2ea00684.png" alt="Screen Shot 2019-12-04 at 4.51.41 PM.png" data-base62-sha1="bjP2KvWFalPHOosA5UarSfw12Oq" width="690" height="203" srcset="https://aws1.discourse-cdn.com/github/optimized/1X/4f5590e1d27491215bc4f1ba99df271f762af756_2_690x203.png, https://aws1.discourse-cdn.com/github/original/1X/4f5590e1d27491215bc4f1ba99df271f762af756.png 1.5x, https://aws1.discourse-cdn.com/github/original/1X/4f5590e1d27491215bc4f1ba99df271f762af756.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/1X/4f5590e1d27491215bc4f1ba99df271f762af756_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screen Shot 2019-12-04 at 4.51.41 PM.png</span><span class="informations">772×228 14.1 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Reply by dukobpa3** [reply]
Should not it be allowed only for users with **force** push access?

**Comment by JamesMudidi** [comment]
<p>This is really a great addition</p>

**Comment by RyanGJones** [comment]
<p>I dont see this aywhere, is this improvement not added in for Github enterprise?</p>

**Comment by siddheshmalpani** [comment]
The beginning of this post has a great suggestion 

> It would be great if I could delete a protected branch as an admin with some kind of confirmation.

The existing feature in protection rules doesn't support that. There could be a use-case where not everyone who has write access to the branch should have delete access. It could be only admins having that access. So if the suggested feature could be implemented, it would be great.

**Comment by alangefe** [comment]
If you do not wish to use the "allow deletions" option for users with push access, another way to get the desired behavior is to rename your branch after it has merged so that it no longer matches the branch protection rules.

For example, if your branch was named "feat/super-cool-stuff" and your protection rule was "feat/*", once you have merged your branch and you are done with it, you can rename it to "merged/feat/super-cool-stuff" so that it no longer matches the branch protection rules, and then you can delete the branch.

**Comment by armory-dci** [comment]
This is good but can we allow only specific team to delete instead everyone who had write access?
